File: 0072-edit-distance/0072-edit-distance.java
class Solution {
    public int min(String word1, String word2, int i, int j, int[][] dp) {
        int n = word1.length();
        int m = word2.length();
        if(i == n){
            return word2.length() - j;
        } 
        if(j == m){
            return word1.length() - i;
        }
        
        if(word1.charAt(i) == word2.charAt(j)){
            return min(word1, word2, i+1, j+1, dp);
        }
        if(dp[i][j] != -1){
            return dp[i][j];
        }

        int insert = min(word1, word2, i, j+1, dp);
        int delete =  min(word1, word2, i+1, j, dp);
        int replace = min(word1, word2, i+1, j+1, dp);

        return  dp[i][j] = 1 + Math.min(insert,Math.min(delete, replace));
    }
    public int minDistance(String word1, String word2) {
        int n = word1.length();
        int m = word2.length();
        int[][] dp = new int[n][m];
        for(int[] d: dp){
            Arrays.fill(d, -1);
        }
        return min(word1, word2, 0, 0, dp);
    }
}

File: 0072-edit-distance/0072-edit-distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int n = word1.length();
        int m = word2.length();
        vector<vector<int>> dp(n+1, vector<int>(m+1, 0));
        for(int i=0;i<=m;i++){
            dp[0][i] = i;
        }
        for(int i=0;i<=n;i++){
            dp[i][0] = i;
        }
        dp[0][0] = 0;
        for(int i=1;i<=n;i++){
            for(int j=1;j<=m;j++){
                if(word1[i-1] == word2[j-1]){
                    dp[i][j] = dp[i-1][j-1];
                }else{
                    int delCharW1 = dp[i-1][j];
                    int repChar = dp[i-1][j-1];
                    int insChar = dp[i][j-1];

                    dp[i][j] = 1 + min({delCharW1, repChar, insChar});
                }
            }
        }
        return dp[n][m];
    }
};